### PR TITLE
"Add support to perform OSD upgrade"

### PR DESCRIPTION
**Pull request by mwaykole**
Signed-off-by: mwaykole <mwaykole@redhat.com>

### CODE REVIEW

**Comment by mwaykole** [comment]
tested full flow http://10.70.37.205:8080/job/rhods-master-job/26/

File: utils/scripts/ocm/ocm.py [hunk 1]
```diff
@@ -74,6 +75,8 @@ def __init__(self, args={}):
         self.osd_major_version = args.get("osd_major_version")
         self.osd_latest_version_data = args.get("osd_latest_version_data")
         self.new_run = args.get("new_run")
+        self.upadte_ocm_channel_json = args.get("upadte_ocm_channel_json")
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:78]: correct the spelling of "update"

File: utils/scripts/ocm/ocm.py [hunk 2]
```diff
@@ -1166,6 +1169,54 @@ def compare_with_old_version_file(self):
...
+    def change_cluster_channel_group(self):
+        """update the channel using ocm cmd"""
+        cluster_id = self.get_osd_cluster_id()
+        run_change_channel_cmd = (
+            "ocm patch /api/clusters_mgmt/v1/clusters/{}"
+            " --body {}".format(cluster_id, self.upadte_ocm_channel_json))
+        log.info(run_change_channel_cmd)
+        ret = execute_command(run_change_channel_cmd)
+        if ret is None:
+            log.info("Failed  to Update the channel")
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1181]: The message can be more specific, Failed to update the channel to <channel_name>

File: utils/scripts/ocm/ocm.py [hunk 3]
```diff
@@ -1166,6 +1169,54 @@ def compare_with_old_version_file(self):
...
+            return ret
+
+
+    def update_ocm_policy(self):
+        """upadte cluster policy to schedule for upgrade osd"""
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1186]: To upgrade OSD ?

File: utils/scripts/ocm/ocm.py [hunk 4]
```diff
@@ -1166,6 +1169,54 @@ def compare_with_old_version_file(self):
...
+        data = read_data_from_json(self.update_policies_json)
+        data["next_run"] = utc_time.replace("\n", "")
+
+        if data["version"] == "latest":
+            get_latest_upgradte_version = (
+                "ocm get cluster {} |"
+                " jq -r '.version.available_upgrades | values'".format(
+                    cluster_id))
+            latest_upgrade_version = execute_command(
+                get_latest_upgradte_version)
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1202]: correct the typo in variable name

File: utils/scripts/ocm/ocm.py [hunk 5]
```diff
@@ -1186,6 +1237,55 @@ def compare_with_old_version_file(self):
...
+        dest="upadte_ocm_channel_json",
+        required=True,
+    )
+    update_ocm_channel.add_argument(
+        "--cluster-name",
+        help="osd cluster name",
+        action="store",
+        dest="cluster_name",
+        metavar="",
+        default="osd-qe-1",
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1284]: Use the default name of cluster that is present in create_cluster.

File: utils/scripts/ocm/ocm.py [hunk 6]
```diff
@@ -1166,6 +1169,54 @@ def compare_with_old_version_file(self):
...
+            data["version"] = latest_upgrade_version
+
+        write_data_in_json(self.update_policies_json, data)
+        schedule_cluster_upgrade = (
+            "ocm post /api/clusters_mgmt/v1/clusters/{}/upgrade_policies"
+            " --body {}".format(cluster_id, self.update_policies_json))
+        ret = execute_command(schedule_cluster_upgrade)
+        if ret is None:
+            log.info("Failed  to Update the Upgrade Policy")
+            return ret
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1217]: Do we have to add a validation checking that the cluster upgrade that is scheduled is correct ?

**Review by aloganat — COMMENTED** [review]
Could u address the sonarcloud failure.

File: utils/scripts/ocm/ocm.py (hunk 6 above)
Comment (mwaykole) [reply utils/scripts/ocm/ocm.py:1217]: the tie we are taking  is from the node of openshift cluster so I guess we don't need it
the give time is in a wrong format or <5 min the ocm patch will fail

File: utils/scripts/ocm/ocm.py (hunk 3 above)
Comment (mwaykole) [reply utils/scripts/ocm/ocm.py:1186]: yes correct it's for openshift dedicated

File: utils/scripts/ocm/ocm.py (hunk 5 above)
Comment (mwaykole) [reply utils/scripts/ocm/ocm.py]: replaced with qeaisrhods-xyz

**Review by tarukumar — DISMISSED** [review]

**Comment by tarukumar** [comment]
@milind i have approved it but please check the black linter action log. it has failures

File: utils/scripts/ocm/ocm.py [hunk 7]
```diff
@@ -1165,6 +1168,50 @@ def compare_with_old_version_file(self):
...
+            latest_upgrade_version = execute_command(get_latest_upgrade_version)
+            log.info(
+                "Version Available to Upgrade are ...{}".format(latest_upgrade_version)
+            )
+            latest_upgrade_version = ast.literal_eval(latest_upgrade_version)[-1]
+            data["version"] = latest_upgrade_version
+        write_data_in_json(self.update_policies_json, data)
+        # fmt: off
+        schedule_cluster_upgrade = (
+            "ocm post /api/clusters_mgmt/v1/clusters/{}/upgrade_policies"\
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1207]: backslash at the end of this line can be removed.

File: utils/scripts/ocm/ocm.py [hunk 8]
```diff
@@ -1186,11 +1233,56 @@ def compare_with_old_version_file(self):
...
     )
+    # Argument of update_ocm_policy
+    update_ocm_policy = subparsers.add_parser(
+        "update_ocm_policy",
+        help="Parser to update_ocm_channel",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+    update_ocm_policy.add_argument(
+        "--update-policies-json",
+        help="pass the json file to update ocm policy",
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1244]: Could you give a example of sample json file here.

File: utils/scripts/ocm/ocm.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import os
```
Comment (aloganat) [inline utils/scripts/ocm/ocm.py:1]: Correct the PR summary to , "Add support to perform OSD upgrade"

File: utils/scripts/ocm/ocm.py (hunk 7 above)
Comment (mwaykole) [reply utils/scripts/ocm/ocm.py:1207]: no @aloganat  arthy the  cmd will fail then 
o/p after removing backslash
``` oc debug node/"$(oc get nodes | awk 'FNR == 2 {print $1}')"
         -- chroot /host date -d '+7 min' -u '+%Y-%m-%dT%H:%M:%SZ' ```

Comment (mwaykole) [reply utils/scripts/ocm/ocm.py:1207]: and the cmd fails in CI if this happens

File: utils/scripts/ocm/ocm.py (hunk 8 above)
Comment (aloganat) [reply utils/scripts/ocm/ocm.py:1244]: Okay, then dont write it here. But, add a sample json file in the downstream repository for reference.

Comment (mwaykole) [reply utils/scripts/ocm/ocm.py:1244]: sure

File: utils/scripts/ocm/ocm.py (hunk 7 above)
Comment (mwaykole) [reply utils/scripts/ocm/ocm.py]: yeah was referring to another cmd  did the changes

**Review by aloganat — APPROVED** [review]
lgtm

**Review by tarukumar — APPROVED** [review]

**Pull request merged**